src/tools/Avalonia.Generators/NameGenerator/AvaloniaNameIncrementalGenerator.cs

```diff
@@ -0,0 +1,204 @@
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using Avalonia.Generators.Common;
+using Avalonia.Generators.Common.Domain;
+using Avalonia.Generators.Compiler;
+using Microsoft.CodeAnalysis;
+using XamlX.Transform;
+
+namespace Avalonia.Generators.NameGenerator;
+
+[Generator(LanguageNames.CSharp)]
+public class AvaloniaNameIncrementalGenerator : IIncrementalGenerator
+{
+    private const string SourceItemGroupMetadata = "build_metadata.AdditionalFiles.SourceItemGroup";
+    private static readonly MiniCompiler s_noopCompiler = MiniCompiler.CreateNoop();
+
+    public void Initialize(IncrementalGeneratorInitializationContext context)
+    {
+        // Map MSBuild properties onto readonly GeneratorOptions.
+        var options = context.AnalyzerConfigOptionsProvider
+            .Select(static (options, _) => new GeneratorOptions(options.GlobalOptions))
+            .WithTrackingName(TrackingNames.XamlGeneratorOptionsProvider);
+
+        // Filter additional texts, we only need Avalonia XAML files.
+        var xamlFiles = context.AdditionalTextsProvider
+            .Combine(options.Combine(context.AnalyzerConfigOptionsProvider))
+            .Where(static pair =>
+            {
+                var text = pair.Left;
+                var (options, optionsProvider) = pair.Right;
+                var filePath = text.Path;
+
+                if (!(filePath.EndsWith(".xaml", StringComparison.OrdinalIgnoreCase) ||
+                      filePath.EndsWith(".paml", StringComparison.OrdinalIgnoreCase) ||
+                      filePath.EndsWith(".axaml", StringComparison.OrdinalIgnoreCase)))
+                {
+                    return false;
+                }
+
+                if (!options.AvaloniaNameGeneratorFilterByPath.Matches(filePath))
+                {
+                    return false;
+                }
+
+                if (!optionsProvider.GetOptions(pair.Left).TryGetValue(SourceItemGroupMetadata, out var itemGroup)
+                    || itemGroup != "AvaloniaXaml")
+                {
+                    return false;
+                }
+
+                return true;
+            })
+            .Select(static (pair, _) => pair.Left)
+            .WithTrackingName(TrackingNames.InputXamlFilesProvider);
+
+        // Actual parsing step. We input XAML files one by one, but don't resolve any types.
+        // That's why we use NoOp type system here, allowing parsing to run detached from C# compilation.
+        // Otherwise we would need to re-parse XAML on any C# file changed.
+        var parsedXamlClasses = xamlFiles
+            .Select(static (file, cancellationToken) =>
+            {
+                var text = file.GetText(cancellationToken);
+                var diagnostics = new List<DiagnosticDescriptor>();
```

Comment:
Nit: consider using `ImmutableArray.CreateBuilder()` here and `diagnostics.MoveToImmutable()` to create the final array.